### PR TITLE
Implement textDocument/signatureHelper.

### PR DESCRIPTION
**Pull request by olafurpg**
This feature is only triggered when opening a parentheses at a function
call-site. The box disappears as soon as you write the first parameter.
We reuse the completion mechanism to find the method symbol and extract
the parameter lists.
 
screen shot 2017-11-22 at 21 49 07

### CODE REVIEW

**Comment by ShaneDelmore** [comment]
I was unable to get this to work.  I tried duplicating your success step exactly but when I type Predef.assert(
then pause I see no dialog, only a message recorded in the output dialog of vscode
```
[Error - 3:29:22 PM] Request completionItem/resolve failed.
  Message: Invalid params
  Code: -32602 
[object Object]
```

tailing metaserver.log shows me a request was made, but I don't see any response for it.  Not sure if I am looking in the correct places or not.
```
15:38:48.997 DEBUG l.core.Connection - Received {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///Users/sdelmore/workspace/language-server/test-workspace/a/src/test/scala/example/UserTest.scala","version":28},"contentChanges":[{"text":"package example\n\nclass UserTest {\n  val user = a.User.apply(name = \"\", 1)\n  user.copy(age = 2)\n  List\n    .apply(1, 2)\n    .map(x => x.+(2))\n  scala.runtime.CharRef.create('a')\n  Predef.assert()\n}\n"}]}}
```

**Comment by laughedelic** [comment]
I just updated my Atom plugin to try this out. And it works!

screen shot 2017-11-23 at 05 00 48

Although I'm not sure it's supposed to look like this.. 🤔 

----

> tailing metaserver.log shows me a request was made, but I don't see any response for it. Not sure if I am looking in the correct places or not.

@ShaneDelmore you see `didChange` notification, but what you should see instead is something like this:

```
DEBUG l.core.Connection - Received {"jsonrpc":"2.0","id":40,"method":"textDocument/signatureHelp","params":{"textDocument":{"uri":"file:///Users/laughedelic/dev/laughedelic/language-server/test-workspace/a/src/main/scala/example/User.scala"},"position":{"line":7,"character":13}}}
```

and response:

```
INFO  s.m.l.Compiler - Signature help at /Users/laughedelic/dev/laughedelic/language-server/test-workspace/a/src/main/scala/example/User.scala:7:13
DEBUG l.core.MessageWriter - payload: {"jsonrpc":"2.0","result":{"isIncomplete":false,"items":[]},"id":39}
DEBUG l.core.MessageWriter - payload: {"jsonrpc":"2.0","result":{"signatures":[{"label":"apply","documentation":"(name: String, age: Int)a.User","parameters":[{"label":"name","documentation":"name: String"},{"label":"age","documentation":"age: Int"}]}]},"id":40}
```

* check that the VS Code extension refers to the right version of the server (now you have to set it explicitly (i.e. `fe14433`))
* check that you opened `test-workspace/a/`.  I opened `test-workspace/` at first and then things were weird, some stuff worked, other not (probably it's Atom specific).

**Comment by laughedelic** [comment]
Here's what I got with some little changes (https://github.com/scalameta/language-server/commit/0749371a49089e38e9239657348e2d5e891e07a2). I think `documentation` field is only for the scaladoc (which we don't have yet, right?), and `label` is directly what will be shown in the interface: 

screen shot 2017-11-23 at 05 38 21

(notice that the first parameter is highlighted in bold)

Probably you guys don't care much for Atom, but I think it's useful to see how different clients implement these things to stay unbiased. Here's [atom-ide-ui docs](https://github.com/facebook-atom/atom-ide-ui/blob/master/docs/signature-help.md#signature-help-wip) for this thing. Some relevant implementation details:

> Once activated, all signature help providers (matching the current grammar scope) will continuously be queried for signatures when the cursor location changes, regardless of the value of triggerCharacters.

**Comment by laughedelic** [comment]
And here is how it looks in VSCode:

screen shot 2017-11-23 at 05 55 23

which is quite the same as in Atom :relaxed: except it let's you to check the second alternative (I guess in atom-ide-ui it's just not implemented yet)

**Comment by gabro** [comment]
Nice! does it highlight the second parameter when you get to it?

**Comment by laughedelic** [comment]
No, but it's just not implemented yet as far as I understand. I didn't read the code in details yet, so I'm not sure how exactly it works.

**Comment by laughedelic** [comment]
Moreover, it does something weird when you start typing:

**Comment by olafurpg** [comment]
@laughedelic that weird behavior is due to the fact that signature helper uses code completions to get the symbols of the methods, it works fine as long as you only invoke it on the trigger character `(`We need to find a more creative way to handle signature help while you fill out the argument list.

**Comment by olafurpg** [comment]
> check that the VS Code extension refers to the right version of the server, that you published locally (now you have to set it explicitly (i.e. fe14433))

This will grow tiring very quickly. Is there any chance to use a stable -SNAPSHOT version number outside of CI?

**Comment by olafurpg** [comment]
@laughedelic I've fixed the weird behavior for basic cases but to make it handle multiple nested arguments we'll need a slightly more sophisticated solution explained in #52. Now it behaves like this

signaturehelp

**Comment by laughedelic** [comment]
> This will grow tiring very quickly. Is there any chance to use a stable -SNAPSHOT version number outside of CI?

I agree. So did you solve it with a9c4cac?

> I've fixed the weird behavior for basic cases

Looks good! 👍

File: metaserver/src/main/scala/scala/meta/languageserver/Compiler.scala [hunk 1]
```diff
@@ -47,41 +54,97 @@ class Compiler(
...
+    getCompiler(path, line, column - 1).fold(noSignatureHelp) {
+      case (compiler, position) =>
+        logger.info(s"Signature help at $path:$line:$column")
+        // Find the last leading open paren to get the method symbol
+        // of this argument list. Note, this may still cause false positives
+        // in cases like `foo(bar(), <cursor>)` since we will find the
+        // symbol of `bar` when we want to match against `foo`.
+        // Related https://github.com/scalameta/language-server/issues/52
+        val lastParenOffset =
+          position.source.content.lastIndexOf('(', position.point)
```
Comment (laughedelic) [inline metaserver/src/main/scala/scala/meta/languageserver/Compiler.scala]: This works outside of parentheses too 😄 

screen shot 2017-11-23 at 16 39 20

File: metaserver/src/main/scala/scala/meta/languageserver/Compiler.scala [hunk 2]
```diff
@@ -47,41 +54,97 @@ class Compiler(
...
+            val parameterInfos = sym.paramLists.headOption.map { params =>
+              params.map { param =>
+                ParameterInformation(
+                  label = s"${param.nameString}: ${param.info.toLongString}",
+                  documentation = None
+                )
+              }
+            }
+            SignatureInformation(
+              label = s"${sym.nameString}${sym.info.toLongString}",
```
Comment (laughedelic) [inline metaserver/src/main/scala/scala/meta/languageserver/Compiler.scala]: I used this just from what already was around. But is there a better way to get separately parameters list and return type?

Comment (laughedelic) [inline metaserver/src/main/scala/scala/meta/languageserver/Compiler.scala]: I used this just from what already was around. But is there a better way to get separately parameters list and return type?

screen shot 2017-11-23 at 13 41 40

File: metaserver/src/main/scala/scala/meta/languageserver/Compiler.scala [hunk 3]
```diff
@@ -47,41 +54,97 @@ class Compiler(
...
+          // bail if we can't find open paren to prevent false positives.
+          noSignatureHelp
+        } else {
+          val lastParenPosition = position.withPoint(lastParenOffset)
+          val signatureInformations = for {
+            member <- compiler
+              .completionsAt(lastParenPosition)
+              .matchingResults()
+              .distinct
+            if member.sym.isMethod
```
Comment (laughedelic) [inline metaserver/src/main/scala/scala/meta/languageserver/Compiler.scala]: What about constructors?

**Comment by olafurpg** [comment]
I think I got it working reasonably well now, it tracks the active argument and doesn't return results anymore if you're outside of an argument list! The activeParameter detection is character based so it will count commas inside string literals so that `foo(", ", <caret>` will say the active argument is 2 when it should be 1. We could use the tokenizer to fix this case, but then we'll have to deal with other problems like unclosed string literals. I think we can leave experiments with that for a future PR!

nov-23-2017 21-23-21

**Comment by olafurpg** [comment]
> What about constructors?

@laughedelic great question! I'll take a look, we can try to fallback to the `apply` member if the symbol we found is not a method

**Comment by olafurpg** [comment]
That `User(<caret>` example @laughedelic crashes :joy:
```scala
java.lang.StringIndexOutOfBoundsException: String index out of range: -2
  java.lang.String.<init>(String.java:196)
  scala.tools.nsc.interactive.Global.typeCompletions$1(Global.scala:1217)
```

**Comment by olafurpg** [comment]
I got `.apply` working too 💃 
screen shot 2017-11-23 at 22 07 20

**Comment by olafurpg** [comment]
Lastly, secondary constructors also get some love ❤️ 

screen shot 2017-11-23 at 22 29 49

Ok, I think this PR is ready to go!

**Comment by gabro** [comment]
Very nice job! These are super useful when paired with autocompletion! A thing I can see improving (for future PRs) is to pretty print the signature. For instance adding a colon between the param list and the result typ

File: metaserver/src/test/scala/tests/compiler/CompilerSuite.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      case (start, end) =>
+        unit.position(start)
+    }
+  }
+
+  trait OverloadHack1; implicit object OverloadHack1 extends OverloadHack1
+  trait OverloadHack2; implicit object OverloadHack2 extends OverloadHack2
+  trait OverloadHack3; implicit object OverloadHack3 extends OverloadHack3
+  trait OverloadHack4; implicit object OverloadHack4 extends OverloadHack4
+  trait OverloadHack5; implicit object OverloadHack5 extends OverloadHack5
```
Comment (gabro) [inline metaserver/src/test/scala/tests/compiler/CompilerSuite.scala]: What's this for? If it's for disambiguating the erased signature, can you maybe use DummyImplicit on just one of the targeted signature below?

**Comment by laughedelic** [comment]
It works now with `User`, but I can't make it work with any constructor from the standard lib, like `List()`, or `sys.process.Process()`. 

Also, this looks strange:
screen shot 2017-11-24 at 02 24 04
screen shot 2017-11-24 at 02 24 23

(besides different typesetting and no syntax highlighting, which is Atom's fault)

It seems that the first parameter list is missing (and brackets around type parameters too)

File: metaserver/src/test/scala/tests/compiler/CompilerSuite.scala (hunk 4 above)
Comment (olafurpg) [reply metaserver/src/test/scala/tests/compiler/CompilerSuite.scala]: This is copied from the scalameta test suite where there are 6 `targeted` overloads. It seems we don't need any overloads here so we can just remove this.

**Comment by olafurpg** [comment]
@laughedelic I got `List(<<1>>` working, that one was tricky since there is both an inferred `.apply` and an inferred type parameter `Int`. I also added special handling for varargs so that the signature help doesn't disappear on the trailing list of arguments.

screen shot 2017-11-24 at 09 21 01
screen shot 2017-11-24 at 09 31 25

Unfortunately, I  couldn't get `Process` to work, the presentation compiler doesn't have the type information. I suspect it's related to how the `.apply` methods a mixed into the `Process` object with a trait.

**Comment by olafurpg** [comment]
@laughedelic I am not able to reproduce the missing argument list for `map`

screen shot 2017-11-24 at 09 45 18

**Comment by olafurpg** [comment]
Gonna merge this since we have #55 using the same testing infrastructure

**Pull request merged**